src/Transposit.ts

```diff
@@ -137,40 +135,85 @@ export class Transposit {
     );
   }
 
+  get stash() {
+    return new Stash(this);
+  }
+
   async loadUser(): Promise<User> {
     this.assertIsSignedIn();
 
-    return await trfetch<User>(this.uri("/api/v1/user"), {
-      method: "GET",
-      headers: {
-        "Content-Type": "application/json",
-        Authorization: `Bearer ${this.accessToken}`,
-      },
-    });
+    return await this.makeCallJson<User>("GET", "/api/v1/user");
   }
 
   async run(
     operationId: string,
     parameters: OperationParameters = {},
   ): Promise<EndRequestLog> {
-    const headers: HeadersInit = {
-      "Content-Type": "application/json",
-    };
-    if (this.accessToken) {
-      headers.Authorization = `Bearer ${this.accessToken}`;
-    }
+    return this.makeCallJson<EndRequestLog>(
+      "POST",
+      `/api/v1/execute/${operationId}`,
+      { body: { parameters } },
+    );
+  }
 
-    return await trfetch<EndRequestLog>(
-      this.uri(`/api/v1/execute/${operationId}`),
-      {
-        method: "POST",
-        headers,
-        body: JSON.stringify({
-          parameters,
-        }),
+  async makeCallJson<T>(
+    method: string,
+    path: string,
+    httpParams: HttpParams = {},
+  ): Promise<T> {
+    const response = await this.makeCall(method, path, httpParams);
+    return response.json().then(
+      x => x,
+      () => {
+        throw new APIError("Failed to read response body", response);
       },
     );
   }
+
+  async makeCall(
+    method: string,
+    path: string,
+    { queryParams, body, headers }: HttpParams = {},
+  ): Promise<Response> {
+    if (headers == null) {
+      headers = {
+        "Content-Type": "application/json",
+      };
+      if (this.accessToken) {
+        headers.Authorization = `Bearer ${this.accessToken}`;
+      }
+    }
+
+    const url = new URL(path, this.hostedAppOrigin);
+    if (queryParams != null) {
+      Object.keys(queryParams).forEach(key =>
+        url.searchParams.append(key, queryParams[key]),
+      );
+    }
+
+    let bodyEncoder;
+    if (headers["Content-Type"] === "application/x-www-form-urlencoded") {
+      bodyEncoder = formUrlEncode;
+    } else {
+      bodyEncoder = JSON.stringify;
+    }
+
+    const response = await trfetch(
+      url.href,
+      Object.assign(
+        { method, headers },
+        body === null ? null : { body: bodyEncoder(body) }, // Only include body if non-null
```

Comment:
I'm a little surprised we are responsible for doing this serialization ourselves. I thought fetch did that for us.

Reply:
We did it beforehand. I can test that fetch does it for us.

Reply:
I tried it out and it doesn't seem to work. I did some digging and someone proposed auto-encoding but they rejected it for introducing too much magic. (And I don't _entirely_ disagree.)

Reply:
yeah, I'm not necessarily asking you to change it as part of this PR, it's just something I happened to notice while reviewing.

Reply:
makes sense, thanks for checking